[FEATURE] Documentation: add `git` guidelines for code-authorities. (#10956)
ILIAS-eLearning/ILIAS: docs/development/git-guidelines.md

# Git Guidelines

> This document defines the git practices we live by inside the ILIAS community when working on this repository. Its

Reviewer: “git” → “Git”. This is a name, and the documentation in the Git project uses the uppercase form.

primary purpose is to ensure transparency, consistency, and clear responsibility when contributing code, reviewing
changes, and integrating them into maintained branches. It also serves as a reference for [authorities who sign off on
code changes](./maintenance.md#authorities).

## 1. Core Principles

All rules and recommendations in this document follow a small set of principles:

> - **Transparency**: Every code change must be traceable to a concrete reason (issue, feature, or release) and to the

Reviewer 2: ```suggestion
- **Transparency**: Every code change must be attributable to a specific cause (issue, feature, or release) and to the
- 
```

Reviewer 2: I think attributable is a more precise English word and I am always mixing up concrete/Beton.

Author: Agree, done.

  discussion that led to its acceptance.

> - **Atomicity**: Each commit must leave the repository in a valid, buildable, and deployable state.

Reviewer 3: I suggest using SHOULD here.

In an ideal world, every commit (and every PR with one or more commits, in rare cases) must leave the code base in a valid state.

I personally see the following problems with a MUST:
- What is the definition of `valid`?
- Do we already have all the tools to guarantee and help developers that a commit fulfills these criteria?
- There might be (again rare) cases where commits or PRs are so big (think about the "Component Revision" and other big refactorings in different components in the last years, even from skilled and experienced developers), which did not leave the system in a `valid` state (at least based on my interpretation of `valid`).

- **Authority**: Changes are reviewed and approved by explicitly defined authorities, and this approval must be visible.

> - **Consistency**: Changes must be applied to all maintained versions of ILIAS by all persons in a controlled and

Reviewer: I do not understand what “all persons” is meant to refer to here.

  reproducible way.

## 2. Commit Types and Semantics

We distinguish between three kinds of commits. The commit type is mandatory and is a first step to set expectations
about its changes.

> - **Bugfix**: Addresses a concrete issue reported in the issue tracker.

Reviewer 2: my proposal would be: "Addresses a specific issue reported in the issue tracker."

> - **Bugfix**: Addresses a concrete issue reported in the issue tracker.

Reviewer 3: One addition: in my opinion, we should introduce an additional type for refactorings. A refactoring commit does not introduce, change, or remove functionality; the behavior remains the same, but the internal structure of the code may be improved or reorganized. These changes also do not necessarily address a reported issue.

Reviewer: I second this.

About bugfixes: we might even want to allow tiny fixes to be done without an associated Mantis issue (e.g. typos spotted while working on something else, things not worth writing an issue for that were reported to us in personal communication – without opening a door for masses of such things to creep in here!).

We might also want to think about how language file changes are classified.

- **Feature**: Introduces, changes, or removes functionality.

> - **Release**: Updates version numbers and dump files.

Reviewer: Just a question: when there is a commit like in pull request #10943, i.e. a commit which updates dependencies, auto-generated files, or the like, is this also a case for type “Release”? Or is this something different?

Author: Good question actually =). I think release should be reserved for actual ILIAS releases. So I would classify them as a feature for now, at least according to my phrasing here.

> Every commit must only be associated to exactly one of these types. Mixing these concerns is not possible.

Reviewer: Side remark: when a new feature adds/modifies some documentation that applies to this feature, I would still happily accept to mark this “feature”.

Author: Agree.

## 3. Commit Message Conventions

> Commit messages must follow a strict structure to ensure readability and transparency. All authorities who sign off on

Reviewer: This mentioned strict structure is not laid out well enough in the following sections. Such a strong formalism is usually favoured because it allows (or increases) machine-readability, and I won’t query that it is a good thing. But the details (spacing, what to do when something has more or less elements, etc.) in the following sections are not detailed enough to help humans write commit messages which are that good. In the most extreme case we would need EBNF or regular expressions; for now I think that adjusting/adding to the explanations and a few examples will do well.

Author: Not exactly sure what I should update here. IMO this structure is clear enough for developers to understand, no?

code changes must ensure they comply with it.

### 3.1 Bugfix

```

> [FIX] #<issue-nr> <component>: <summary>

Reviewer 2: I was wondering what a good length for a commit title might be. I read often something like 50 characters (https://cbea.ms/git-commit/) but I think this would only apply to your "summary" proposal. What do you think?

Reviewer 3: We also refer to https://cbea.ms/git-commit/ in our internal Git guidelines, so I support the idea of using 50 as a rule of thumb.

Reviewer: The length of 50 characters probably originates from use in mail subjects. It is utilised in various Git tools and platforms. Even more authoritative sources have it. I agree that this is the most sensible length for the first line of a commit message.

When we introduce prefixes, suffixes or any other extensions to this line, we may still decide to keep the 50-character limit _for the rest_ (i.e. `<summary>`), which I can accept. We should nevertheless try to make static values (e.g. for marking types) short so that precious space is not wasted, as lines exceeding the 50-character limit may be abbreviated and we like to minimise that.

Author: I have added this as rule of thumb to each `<summary>` placeholder description.

<description>
```

> * `<issue-nr>`: the issue number being addressed.

Reviewer: Correct English abbreviation for “number” is “no.” (“Nr.” is German). So I recommend `<issue-no>`.

> * `<component>`: the ILIAS component responsible for the affected files. If multiple components are involved, list them

Reviewer: With this explanation, I have trouble locating the component to pick. Is it about the `components/` directory in the ILIAS tree or is it a reference to `docs/development/maintenance.md`? This needs clarification.

Hint: the explanation is good when developers are able to compile a list of valid component identifiers, and when they are also able to effortlessly select the proper one in almost every case.

Author: I tried to clarify.

>   comma-separated. Responsibility is determined by ownership, not by syntactic location (e.g. `Setup` owns

Reviewer 2: to be honest: I don't understand what you want to say here. Could you explain this?

>   comma-separated. Responsibility is determined by ownership, not by syntactic location (e.g. `Setup` owns

Reviewer: Separated by comma or by comma+space? (I would accept both. And I would prefer the latter.) This needs clarification to ensure consistency.

Author: I actually switched to `/` and no spaces. I think this `,` without space would be weird, but the space is also kind of unnecessary. `/` is IMO more suitable.

  `./cli/setup.php`, `Language` owns `./lang`).
* `<summary>`: a concise, factual description of the change.

> * `<description>` (optional): further details, typically an unordered list.

Reviewer: I cannot understand the last part and strongly suggest to remove “, typically an unordered list”.

_Maybe_ this is related: I have already seen commits in our Git history that contain a bulleted list with some changes. In my experience this was either a result of a lot of changes mixed into one single commit (very, very, bad practice! See below for more about this) or just a way of conserving words.

Code and commit messages are read many more times than written. Developers should really invest that bit of extra time to explain why their commit is structured like it is, and what it is about, possibly give further information, maybe add references, and so on, so that others can well follow. I see no reason why we should mislead developers who would otherwise have written longer explanations to instead present unordered lists which may be difficult to understand.

Author: True, totally agree. I removed that part.

### 3.2 Feature

```
[FEATURE] <component>: <summary>

<description>
```

> * `<component>`: the component implementing the feature. Multiple components must be listed comma-separated. If only a

Reviewer: See my comment above about separation by comma or comma+space. Same here; clarify.

  usage is adapted, this component may be omitted.
* `<summary>`: a concise description of the functional change.

> * `<description>`: a detailed explanation of the implementation or behaviour, ideally linking the feature wiki entry.

Reviewer: Not just ideally, is it?

What about:
> A commit introducing a feature SHOULD provide a link to the corresponding page in the feature wiki (if such page exists).

Links to other relevant information are nice as well.

> ### 3.3 Release

Reviewer 2: @fwolf-ilias  is this good for you?

```
[RELEASE] <version>
```

> * `<version>`: the release version, prefixed with `v`, mirroring the associated git tag.

Reviewer: “git” → “Git” (see above).

## 3.4 Components and Component Substitutes

> Most commits can be attributed to a concrete ILIAS component. For other locations inside the repository without a clear

Reviewer 2: "concrete" -> "specific"

> component associated with it, the following substitutes must be used:

Reviewer 2: Please provide an example what this substitute might look like.

> * `Documentation`: changes to `./docs` and other top-level documentation files (e.g. `README.md`, `LICENSE`).

Reviewer: While indeed `LICENSE` is the only other file that currently matches in the repository root, should we really handle licensing changes under a “documentation” mark? This might be dangerous in that readers overlook them, assuming them to be just “unimportant documentation”. On the other hand, we can argue that it is now specified and readers should better be aware of the possibility.

Author: I agree, however, I don't think we're ever going to change the license and only communicate this using a commit. This will certainly be noticed, even if we categorise it as a documentation change. I therefore leave it for the time being.

* `CI`: changes to automation, QA tooling, scripts, or related configuration (e.g. `./scripts`, `./.github`,
  `./captainhook.json`).
* `Dependency`: changes to `composer` or `npm` related files. These files must not be edited manually and usually change
  only due to dependency updates.

> * `Authority`: changes to `./docs/development/maintenance.md` that affect authority definitions only.

Reviewer: I suggest to shorten these substitute names, for the reasons stated above.

`doc`, `CI`, `dep` sound nice. We should _not_ use `auth`, because this is usually understood as “authentication”. `autho` does not sound nice, but may be better. I’m open to improvement suggestions.

Author: I personally dislike abbreviations very much, especially if they are not known acronyms. This becomes ambiguous really fast (dep could also be department, doc could also be a doctor). CI is the only known acronym in this case, so I vote to spell out everything else.

_If none of these substitutes match the changes of the commit, this document may be extended and the component or

> substitute may be omitted for the time being._

Reviewer: I suggest a rewording here: primary effect is that the omission is accepted for the time being. Secondary is that this document may be updated.

What about:
> _If none of these substitutes match the changes of the commit, the component or
> substitute may be omitted for the time being. This document may be extended in
> future to take more cases into account._

## 3.5 Commits Originating from Pull Requests

> Commits that originate from pull requests must preserve a reference to the GitHub discussion. This ensures that design

Reviewer 2: I think it's called "conversation", not "discussion". Discussion is imo another github feature

Reviewer: Maybe what was meant is the discussion in the conversation tab/section on GitHub. :smile: But that may be too wordy; Rob’s suggestion is fine.

decisions and review outcomes remain easily traceable and accessible.

> So when integrating a pull request, the final commit message must be amended as follows:

Reviewer: Are we really sure we like maintainers to reword commit messages? (Git uses the term “reword” when commit messages are changed; this is a sub-operation of amending a commit.)

This will obviously break signed commits. Though when using/allowing signed commits we have other unhandled problems, so we may accept this for now.

We can ask contributors to use that syntax right from the start, which may help.

I prefer to put a reference (may also be a URI) to the pull request in the body of the commit message because committers can easily produce this for commits for which it fits, and they can decide into which of their commits it should be put. This does not solve the problem of a possible necessary merge commit though, which would be made by the maintainer (we’ll get to merging later).

Author: I updated this section, I now only say that it must be correct in the end. Who rectifies it can be decided by the authorities.

```
<commit-message> (#<pr-number>)
```

* `<commit-message>`: a commit message that complies to the conventions above.
* `<pr-number>`: the pull request number as shown in GitHub.

**This appendix is mandatory for all PR-based commits.** GitHub will automatically create a link to the pull request and
mention cherry-picked commits with the same appendix inside the pull requests timeline, leaving a transparent overview.

> ## 6. Reviewing Pull Requests

Reviewer 2: I think the numbering started to be off with "3.4 Components and Component Substitutes"

Pull requests are reviewed using GitHub's built-in review features. Reviews by authorities must be visible in the PR
timeline and explicitly signal approval, rejection, or required changes.

> Every pull request that is integrated into the repository should, ideally, be explicitly approved by the responsible

Reviewer 2: I don't think we need "ideally" here. Removing it will provide a clearer sentence.

Reviewer: Furthermore, we may like to turn the formulation around: when a pull request is expressly approved by the authority in charge, it may be integrated (merged). (And such integration happens by said authority.)

authority.

### 6.1 Review Structure

Reviews should be structured and grouped into questions, suggestions, and change requests. Reviewers are encouraged to
use the following template:

```markdown
Hi @<contributor>,

> Thx a lot for your contribution to <component-name>!

Reviewer: Can we please write “thank you” or similar, and not suggest “thx”? Thanks. :smile:

<general-feedback>

> About the concrete changes, please answer the following questions:

Reviewer 2: I'd prefer a more defensive approach. This could be: "Please help me out in answering the following questions:".

- [ ] <reference>: <question>?

Please consider the following suggestions. You do not need to follow them, but please indicate briefly why you prefer to

> do otherwise:

Reviewer: Suggestion for the second sentence:
> If you decide not to follow them, please indicate your reason(s) briefly.

- [ ] <reference>: <suggestion>

Please implement the following changes:

- [ ] <reference>: <change-request>

> 

Reviewer: Shouldn’t we sort from most important to less important? That is, swap the preceding two sections so that change requests precede suggestions? 🤷

Author: I personally think its good to have this order, because otherwise you might just skim the less important parts. But I guess at the end its simply preference.

Since we have used this for some time inside the UI framework now, I will leave it like that.

Kind regards,
@<member> (as <component-name> <authority>)
```

The placeholders inside the template must be replaced by the following values, if necessary:

* `<contributor>`: GitHub username of the person who opened the pull request.
* `<general-feedback>`: general feedback on the proposed changes that does not fit into any of the specific categories,
  such as background information, context, or relevant state of affairs.
* `<reference>`: a concrete anchor for a review point that allows the contributor to respond precisely and enter a focused
  discussion. This is typically a class name, object, method, file name, or a short descriptive identifier.

>   class-, object- or filename, or other few words that can be quickly referenced.

Reviewer 2: ```suggestion
  Class-, object- or filename, or other few words that can be quickly referenced.
```

Reviewer: Is this line maybe a duplicate from an earlier rewording. Can we just remove it?

* `<question>`: a concrete question about the changes that must be answered before the pull request can be integrated.
* `<suggestion>`: a concrete improvement proposal that requires either implementation or explicit feedback from the
  contributor before integration.
* `<change-request>`: a concrete change that must be implemented by the contributor before the pull request can be
  integrated.
* `<member>`: GitHub username of the person reviewing the pull request.
* `<component-name>`: name of the ILIAS component(s) the reviewer has authority over.

> * `<authority>`: tthe authority under which the review is performed (usually code).

Reviewer 2: ```suggestion
* `<authority>`: the authority under which the review is performed (usually code).
```

Sections of the template may be omitted if they are not applicable. For trivial or spotless changes, a minimal review or
direct approval is acceptable, as long as the signature clearly indicates who signed off on the changes and for which

> components.

Reviewer: This sounds like an obvious candidate for `Signed-off-by` trailers and/or formalising what the committer means in this case. The former is more flexible. We can revisit this in future.

Author: Agree, I assume your iteration might cover this =).

If no clear authority exists for the affected code, the pull request falls into the shepherd model. In this case, the
signature must reflect this by using "as shepherd" instead of a component-specific authority.

### 6.2 Review Outcomes

GitHub distinguishes between three review outcomes, which we interpret as follows:

- **Comment**: no required changes or suggestions, but open questions.
- **Request Changes**: at least one suggestion or change request must be addressed before integration.
- **Approve**: the pull request is ready to be integrated or all review points have been resolved.

## 7. Integrating Pull Requests

> All pull requests must be integrated using the **squash and merge** option. This ensures a clean history and enforces

Reviewer 3: Again: I suggest using a SHOULD here.

I like the idea of having atomic commits, but I think it might occur (in rare cases) that a PR consists of multiple atomic commits.

Reviewer: I strongly object to the use of squashing for the integration of contributions. I’m going to detail this further in short time; for now, this is a summary:

* makes reading code _much_ harder (tooling like `git log`, `git blame` and GUI tools impacted),
* breaks signed commits (obviously),
* needs _very_ careful handling by integrators not to violate license: contributions licensed under GPL-3.0-only or GPL-3.0-or-later require attribution; in case of several commits with various authors the committer (=authority, =person who does the merge) needs to carefully insert `Co-authored-by` trailers or otherwise ensure that the attribution is not lost.

Author: Agree, done.

atomic commits to the repository. Squashing the commits also gives us the opportunity to amend the commit message, to

> make sure they all comply with our conventions. **For traceability its important to add the

Reviewer 2: I prefer "attributability" over "traceability"

[PR appendix](#35-commits-originating-from-pull-requests) during this step**.

As a general rule of thumb, the following rules apply:

* Prefer a single commit per pull request.
* Multiple commits are acceptable only if they represent clearly separable concerns. However, this will prevent adding
  the PR appendix to the commit message, so an according change must be requested by the contributor.
* The final commit message(s) must comply with the defined conventions.

> * The commit must include a PR reference appendix.

Reviewer: These rules sound like we try have the merging authorities rectify contributors’ bad practices. I’d rather say we solve this by asking contributors to overdo their work when necessary and finally merge the state when the authority is happy. This removes some burden from authorities and educates contributors at the same time. Overall it will increase code quality and provide more (publicly readable) requests to fix this and that; these again can help future contributors.

We should all be aware that having a single commit for a potentially huge feature is a big burden for the entire community working on the code, and _will_ get back at us in future when someone needs to wade through that giant block when there is some problem that needs debugging or when someone likes to implement something similar or wants to look something up.

Author: Agree, done. I simply say that integrated commits must also adhere to the guideline now, keeping it flexible on who ultimately does the fixing.

* Integration must be performed by an authority responsible for the affected component(s). Integration also includes

>   the integration into other relevant branches of the repository.

Reviewer: Just to make sure I understand this correctly: does this mean that nobody but the authority in charge is allowed to do integrations (on any branch)? Or is it meant to say that an authority is obliged to merge something to affected-branch-B, too, after they have decided to merge it to affected-branch-A? The former is obvious (and _might_ be slightly extended to allow for security-related overrides or helping another authority get things done or allow the Product Manager to merge something, etc.). The latter would mean that a complicated transfer to a (very different) new version would have to be priced into the work _before_ beginning the merge to branch A.

I think the former is what is meant, isn’t it? Maybe we can improve the wording here.

Author: The former was meant, but I dropped this section anyways now. I think this document is vague enough about who is permitted to perform integrations so that the document on authorities could define this more precisely.

### 7.1 Recommended Integration Procedure

1. Review the pull request.
2. Inspect the commit history and ensure all commits can be squashed.
3. Determine the target branches for integration.
4. Perform a squash merge and edit the commit message:

>     * ensure compliance with commit message conventions;

Reviewer 2: This bullet points ends with ";" the next one with "." This is not clear imo.

Reviewer: Correct in de-DE for lists; I think this is okay for en-UK, too (but let’s ask language experts if unsure).

    * append the PR number.
5. Finalise the merge.

> 6. Cherry-pick the resulting commit into all relevant maintained branches.

Reviewer: Cherry-pick a single commit, created by squashing several others, to other branches? Good luck. (In practice, authorities will almost always rebase the commit chain and later squash the commits in order to get the same result. Everything else is far too much hassle.)

Author: I removed the recommended procedure. Its partially covered by the new section on how commits are integrated in general.

## 8. Cherry-picking Commits

To apply changes across different major versions of ILIAS, we use `git cherry-pick`. This is all straightforward, except
for changes that affect compiled assets (e.g. CSS or bundled JavaScript), that are tracked inside the repository.

> When cherry-picking commits that affect such resources:

Reviewer: Nah, rather make sure that compilation of assets happens in separate commits, then re-run the compilation tools on another branch and we’re fine. Much easier and also easier to work with for people receiving the code.

Author: Agree, done.
